File: 03-herencia/java/Moto.java
/*
 * Moto.java - OTRA CLASE HIJA
 *
 * Hereda de lo mismo que Auto, pero resuelve describir() a su manera.
 * Ese es el germen del polimorfismo (demo 04).
 */
public class Moto extends Vehiculo {

    private boolean tieneCarenado;

    public Moto(String modelo, int anio, boolean tieneCarenado) {
        super(modelo, anio);
        this.tieneCarenado = tieneCarenado;
    }

    @Override
    public void describir() {
        String extra = this.tieneCarenado ? "con carenado" : "sin carenado";
        System.out.println("Moto " + this.modelo + " (" + this.anio + "), " + extra);
    }

    public void hacerCaballito() {
        System.out.println(this.modelo + ": caballito!");
    }
}

File: 03-herencia/python/moto.py
"""
moto.py - OTRA CLASE HIJA

OJO: en Python no existe @Override. Si escribes "descrivir" en vez de
"describir", Python crea un metodo NUEVO en silencio y nunca te enteras:
el objeto seguira usando el describir() del padre.
En Java, @Override hace que eso no compile.
"""

from vehiculo import Vehiculo


class Moto(Vehiculo):

    def __init__(self, modelo, anio, tiene_carenado):
        super().__init__(modelo, anio)
        self._tiene_carenado = tiene_carenado

    def describir(self):
        extra = "con carenado" if self._tiene_carenado else "sin carenado"
        print(f"Moto {self._modelo} ({self._anio}), {extra}")

    def hacer_caballito(self):
        print(f"{self._modelo}: caballito!")
